Treat "lookup into immediately imported modules" as a special case

jrose-apple opened the pull request:
Features like `@testable import` change the results produced by name lookup in the source file with the testable import. What they *shouldn't* do is change the results of lookup into the same module from elsewhere in the import graph...and neither should cached results from lookup elsewhere in the import graph affect the lookups from the source file. Encode this difference in the cache used during module-level name lookup to fix testable imports like this.

(The test case here looks a little contrived because of `@_exported`, but that's how imports are treated in Clang modules and with bridging headers, so it is in fact a realistic scenario.)

rdar://problem/48890959

jrose-apple committed:
Don't create import loops in mixed-source frameworks w/ submodules

Swift handles Clang submodules as far as controlling visibility, but
pretends that all the declarations live in the top-level module for
the purposes of lookup. This was interacting poorly with another
feature: automatically importing the Swift part of a mixed-source
framework and using that as the "presentation" of a Clang module...
or a submodule in the same framework. Therefore, if you are

- compiling the Swift part of a mixed-source framework, and
- importing one of the framework's (Clang) submodules

name binding would say that the current file imports the Swift module
currently being compiled, in addition to the Clang module we actually
want. Just break the cycle in that case.

jrose-apple committed:
Treat "lookup into immediately imported modules" as a special case

Features like `@testable import` change the results produced by name
lookup in the source file with the testable import. What they
/shouldn't/ do is change the results of lookup into the same module
from elsewhere in the import graph...and neither should cached results
from lookup elsewhere in the import graph affect the lookups from the
source file. Encode this difference in the cache used during
module-level name lookup to fix testable imports like this.

(The test case here looks a little contrived because of '@_exported',
but that's how imports are treated in Clang modules and with bridging
headers, so it is in fact a realistic scenario.)

rdar://problem/48890959

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@davidungar will rightfully say that ModuleNameLookup.cpp is too complicated, and it is, but I don't think I've made it *more* complicated.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -174,7 +178,10 @@ static void lookupInModule(ModuleDecl *module, ModuleDecl::AccessPathTy accessPa
 
-    // This only applies to immediate imports of the top-level module.
+    // Special treatment based on the use site only applies to immediate 
+    // imports of the top-level module.
```
Why does it only apply in some cases? Could the comment explain that?

davidungar reviewed (commented):
As you anticipated, I found this code, the ~100-line `lookupInModule` function to be quite inscrutable, even beyond merely complicated. It is long, there are not many comments, and there is mutable state. Since you understand this code, you could do future team-members a great service by fixing it up. Do the time pressures you are under permit it?

jrose-apple committed:
Refactor module name lookup with a helper class

Break up the large 'lookupInModule' function into a class with helper
methods; eliminate parameters that can be inferred from context or
stored as fields; add more comments and 'const'.

No intended functionality change.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@davidungar Finally picked this up again, this time with (a form of) the refactoring you asked for.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -19,9 +19,12 @@ using namespace swift;
-                                                TinyPtrVector<ValueDecl *>,
-                                                32>;
+  // The usable results of a lookup in a particular module may differ based on
+  // where the lookup is happening.
```
It's not immediately apparent what you are thinking of,  because the "where" could be more specific. Or maybe it's the "usable"? Do you any desire to expound a bit? Maybe with an example?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -19,9 +19,12 @@ using namespace swift;
+  using ModuleLookupCacheKey = std::pair<ModuleDecl::ImportedModule,
+                                         const DeclContext * /*lookupScope*/>;
+  using ModuleLookupCache = llvm::SmallDenseMap<ModuleLookupCacheKey,
+                                                TinyPtrVector<ValueDecl *>, 32>;
```
I've always wondered where the sizes of the Small collections come from. Is 32 just a guess, or is there some performance cliff if it were to change to 16? It's one of those small distractions in reading the code for me which can add up over time. Is the expectation that most apps have < 32 modules? Are there measurements?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -40,6 +43,8 @@ namespace {
 /// Returns true if this particular ValueDecl is overloadable.
 static bool isOverloadable(const ValueDecl *VD) {
+  // FIXME: This is very suspect; it doesn't really match how the rest of the
+  // compiler works anymore. How do we want cross-module shadowing to work?
```
Would it be possible to point to a document or some code that shows how the rest of the compiler does work? Or say something about what is meant?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a CanTypeSet to resolve overloads, and have
+/// the given \p resolutionKind.
+static bool canReturnEarly(const CanTypeSet &overloads,
```
I'm confused about the `overloads` argument here. It doesn't seem to be used, and this method is not polymorphic. What's the reason for passing it in?
Ah--after looking at the whole file, I get it: it's statically overloading on the type of that argument.
And you juxtaposed them to make that clear.
What would probably help is a bit more "why" to go with the what. Is there a comment somewhere about why we need a Named wobulator someplaces vs a vanilla one other places?

jrose-apple replied:
Yeah, I thought about this some more and this should all just be wrapped up in a strategy type.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a NamedCanTypeSet to resolve overloads, and
+/// have the given \p resolutionKind.
```
My own style would be comment it by saying something like: "This function statically overrides the previous one to ensure that the result is always false for this argument type." But even that seems not "why" enough...

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+}
+
+namespace {
+using LookupCallbackTy = llvm::function_ref<
```
At this point in reading the code, I'm wondering "Why do we need a callback for lookups?"

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+  bool recordImportDecls(SmallVectorImpl<ValueDecl *> &results,
+                         ArrayRef<ValueDecl *> newDecls,
+                         OverloadSetTy &overloads,
+                         ResolutionKind &resolutionKind);
```
Is resolutionKind just an input parameter? If so, could it be `const`?

jrose-apple replied:
No, it's in/out. I may not use `const` on pass-by-value parameters, but I always do it for references.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+  /// Given a list of imports and an access path to limit by, perform a lookup
+  /// into each of them and record the results in \p decls, filtering based on
+  /// the given resolution kind and existing overload set.
+  void collectLookupResultsFromImports(
```
Again, could the input refs be const refs?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+public:
+  /// Performs a qualified lookup into the given module and, if necessary, its
+  /// reexports, observing proper shadowing rules.
+  static void lookupInModule(ModuleDecl *startModule,
```
At this point, seeing approx the same bunch o' parameters a few times, I'm starting to think about a missing concept--just a thought.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
   switch (resolutionKind) {
   case ResolutionKind::Overloadable: {
     // Add new decls if they provide a new overload. Note that the new decls
     // may be ambiguous with respect to each other, just not any existing decls.
```
I realize that this wording is old, but does "any existing decls" mean any anywhere? Or just in one of these data structures here?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
     // may be ambiguous with respect to each other, just not any existing decls.
-    std::copy_if(newDecls.begin(), newDecls.end(), std::back_inserter(results),
-                 [&](ValueDecl *result) -> bool {
+    llvm::copy_if(newDecls, std::back_inserter(results),
```
Nice!

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
   }
 
   case ResolutionKind::Exact:
     // Add all decls. If they're ambiguous, they're ambiguous.
```
Again, the second sentence seems like it's trying to say more than it does. Could you take a whack at a clarification?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+}
+
+template <typename OverloadSetTy>
+void ModuleNameLookup<OverloadSetTy>::collectLookupResultsFromImports(
```
Nice expressive function name!

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+    const DeclContext *moduleScopeContext,
+    OverloadSetTy &overloads, ResolutionKind resolutionKind) {
+
+  // Prefer scoped imports (import func Swift.max) to whole-module imports.
```
I like the example, but how could `Swift.max` ever be mistaken for a whole-module import?

jrose-apple replied:
Clearly the comment is not clear enough. :-)

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+template <typename OverloadSetTy>
+void ModuleNameLookup<OverloadSetTy>::collectLookupResultsFromImports(
+    ArrayRef<ModuleDecl::ImportedModule> reexports,
+    ModuleDecl::AccessPathTy accessPath,
```
`const` for accessPath?

jrose-apple replied:
>< I'm still not in the habit of doing that for my parameters, but sure.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+void ModuleNameLookup<OverloadSetTy>::collectLookupResultsFromImports(
+    ArrayRef<ModuleDecl::ImportedModule> reexports,
+    ModuleDecl::AccessPathTy accessPath,
+    SmallVectorImpl<ValueDecl *> &decls,
```
Is `decls` an out parameter?

jrose-apple replied:
Yep. Does it need something explicit beyond being a mutable reference?

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+    OverloadSetTy &overloads, ResolutionKind resolutionKind) {
+
+  // Prefer scoped imports (import func Swift.max) to whole-module imports.
+  SmallVector<ValueDecl *, 8> unscopedValues;
```
8? I realize that this is an llvm idiom. There's an opportunity to add clarity with something like
```
const size_t likelyMaxImportsInAFile = 8
SmallVector<..., likelyMaxImportsInAFile> ...
```
or whatever this number means.

jrose-apple replied:
Yeah. :-/ It's true that we just pick a number these days, but I'm not going to try to justify this one right now.

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+    } else if (!next.first.empty() &&
+               !ModuleDecl::isSameAccessPath(next.first, accessPath)) {
+      // If we ever allow importing non-top-level decls, it's possible the
+      // rule above isn't what we want.
```
Great comment!

davidungar commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -118,130 +214,147 @@ static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
+    SmallVectorImpl<ValueDecl *> &decls,
+    ResolutionKind initialResolutionKind,
+    const DeclContext *moduleScopeContext,
+    ArrayRef<ModuleDecl::ImportedModule> extraImports) {
```
from 11 params to 8--that's an improvement!

davidungar reviewed (approved):
The refactoring is nice! I gave it a shallow look--can't vouch for its correctness. (If you want that, let me know and I'll take the time to study it.) But I made a few suggestions about where a clarification could help. Feel free to pick and choose.

jrose-apple commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -18,93 +18,263 @@
+      OverloadSetEntry &entry = overloads[result->getBaseName()];
+      if (!isOverloadable(result)) {
+        entry.first = ResolutionKind::Exact;
+        entry.second.clear();
```
This `clear()` is the one new bit of behavior that got folded into this commit. It might mildly reduce memory usage during these lookups.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
@swift-ci Please smoke test
